tools/istio-iptables/pkg/capture/run.go

```diff
@@ -754,38 +754,219 @@ func (cfg *IptablesConfigurator) executeIptablesCommands(iptVer *dep.IptablesVer
 	return nil
 }
 
-func (cfg *IptablesConfigurator) executeIptablesRestoreCommand(iptVer *dep.IptablesVersion, isIpv4 bool) error {
-	var data string
-	if isIpv4 {
-		data = cfg.ruleBuilder.BuildV4Restore()
-	} else {
-		data = cfg.ruleBuilder.BuildV6Restore()
+func (cfg *IptablesConfigurator) tryExecuteIptablesCommands(iptVer *dep.IptablesVersion, commands [][]string) {
+	for _, cmd := range commands {
+		cfg.ext.RunQuietlyAndIgnore(constants.IPTables, iptVer, nil, cmd...)
 	}
+}
 
+func (cfg *IptablesConfigurator) executeIptablesRestoreCommand(iptVer *dep.IptablesVersion, data string) error {
 	log.Infof("Running iptables restore with: %s and the following input:\n%v", iptVer.CmdToString(constants.IPTablesRestore), strings.TrimSpace(data))
 	// --noflush to prevent flushing/deleting previous contents from table
 	return cfg.ext.Run(constants.IPTablesRestore, iptVer, strings.NewReader(data), "--noflush")
 }
 
-func (cfg *IptablesConfigurator) executeCommands(iptVer, ipt6Ver *dep.IptablesVersion) error {
-	if cfg.cfg.RestoreFormat {
-		// Execute iptables-restore
-		if err := cfg.executeIptablesRestoreCommand(iptVer, true); err != nil {
-			return err
+// getStateFromSave function takes a string in iptables-restore format and returns a map of the tables, chains, and rules.
+// Note that if this function is used to parse iptables-save output, the rules may have changed since they were first applied
+// as rules do not necessarily undergo a round-trip through the kernel in the same form.
+// Therefore, these rules should not be used for any critical checks.
+func (cfg *IptablesConfigurator) getStateFromSave(data string) map[string]map[string][]string {
+	lines := strings.Split(data, "\n")
+	result := make(map[string]map[string][]string)
+	for _, defaultTable := range []string{constants.FILTER, constants.NAT, constants.MANGLE, constants.RAW} {
+		result[defaultTable] = make(map[string][]string)
+	}
+
+	table := ""
+	chain := ""
+	for _, line := range lines {
+		if strings.TrimSpace(line) == "" {
+			continue
 		}
-		// Execute ip6tables-restore
-		if err := cfg.executeIptablesRestoreCommand(ipt6Ver, false); err != nil {
-			return err
+		if strings.HasPrefix(line, "#") || line == "COMMIT" {
+			continue
+		}
+		// Found table
+		if strings.HasPrefix(line, "*") {
+			chain = ""
+			table = strings.TrimSpace(line[1:])
+			continue
+		}
+		// Found chain, setup an empty list for the chain if it is an ISTIO one
+		if strings.HasPrefix(line, ":") {
+			if !strings.HasPrefix(line, ":ISTIO") {
+				continue
+			}
+			chain := strings.Split(line, " ")[0][1:]
+			_, ok := result[table][chain]
+			if !ok {
+				result[table][chain] = []string{}
+			}
+			continue
+		}
+
+		rule := strings.Split(line, " ")
+		ruleChain := ""
+		if chain == "" {
```

Comment:
This is broken I think. `chain` is always `""` since L800 is ineffective. This implies we don't have sufficient testing

Reply:
This is dead-code left over from previous iterations of the PR.
The chain value at the loop level should be removed because it didn't work as intended. The 'chain' variable is incorrect because having :CHAIN in the iptables-save output doesn't guarantee that all subsequent rules (at least until the next :CHAIN) belong to that chain.

The tests didn't fail because this variable is always ignored in favor of the correct 'ruleChain' variable. I will clean up the variable and remove the useless if condition.

Reply:
Removed